p2panda-js/src/session.ts

```diff
@@ -0,0 +1,278 @@
+
+const log = debug('p2panda-js:session');
+
+export default class Session {
+  // Address of a p2panda node that we can connect to
+  endpoint: string;
+
+  // An rpc client connected to the confgiured endpoint
+  client: Client;
+
+  // The wasm library from p2panda-rs. To ensure that it is loaded before
+  // using it await `this.loadWasm()`
+  // @ts-expect-error p2panda must be resolved before accessing it
+  p2panda: Resolved<typeof p2panda>;
+
+  // Cached arguments for the next entry
+  nextEntryArgs: { [cacheKey: string]: EntryArgs } = {};
+
+  constructor(endpoint: Session['endpoint']) {
+    this.endpoint = endpoint;
+    const transport = new HTTPTransport(endpoint);
+    this.client = new Client(new RequestManager([transport]));
+  }
+
+  __schema: string | null = null;
+
+  /**
+   * Return currently configured schema.
+   *
+   * Throws if no schema is configured.
+   */
+  get _schema(): string {
+    if (!this.__schema) {
+      throw new Error(
+        'Requires a schema. Configure a schema with ' +
+          '`session.schema()` or with the `options` parameter on methods.',
+      );
+    }
+    return this._schema;
+  }
+
+  set _schema(val: string) {
+    this.__schema = val;
+  }
+
+  /**
+   * Set a fixed schema for this session, which will be used if no other schema
+   * is defined through a methods `options` parameter.
+   *
+   * @param val schema hash
+   * @returns Session
+   */
+  schema(val: string): Session {
+    this._schema = val;
+    return this;
+  }
+
+  __keyPair: KeyPair | null = null;
+
+  get _keyPair(): KeyPair {
+    if (!this.__keyPair) {
+      throw new Error(
+        'Requires a signing key pair. Configure a key pair with ' +
+          '`session.keyPair()` or with the `options` parameter on methods.',
+      );
+    }
+    return this._keyPair;
+  }
+
+  set _keyPair(val: KeyPair) {
+    this.__keyPair = val;
+  }
+
+  /**
+   * Set a fixed key pair for this session, which will be used by methods unless
+   * a different key pair is configured through their `options` parameters.
+   *
+   * @param val key pair instance generated using the `KeyPair` class.
+   * @returns key pair instance
+   */
+  keyPair(val: KeyPair): Session {
+    this._keyPair = val;
+    return this;
+  }
+
+  // Load and return the WebAssembly p2panda library.
+  //
+  // Always await this function before using `this.p2panda`. Unfortunately this
+  // cannot be handled in the constructor as the contructor cannot be async.
+  async loadWasm(): Promise<P2Panda> {
+    if (this.p2panda == null) {
+      this.p2panda = await p2panda;
+
+      // I am removing this again because it breaks the node build but it may
+      // have to go back in.
+
+      // if (this.p2panda.default != null) {
+      //   log('fallback loader from `p2panda.default`');
+      //   // @ts-expect-error only applies for node context
+      //   this.p2panda = await p2panda.default;
+      // } else {
+      //   log('loaded wasm lib');
+      // }
```

Comment:
Dead code?

Reply:
I'd suggest making an issue instead and removing this :+1:

Reply:
I don't remember exactly what the issue actually was. I think it was that we are only testing this library in a node environment with our test suite and there was a problem with the bundling in browser environments. I will make an issue to test that.

Reply:
https://github.com/p2panda/p2panda/issues/82